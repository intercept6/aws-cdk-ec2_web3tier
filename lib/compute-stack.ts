import cdk = require('@aws-cdk/core');
import {AutoScalingGroup} from "@aws-cdk/aws-autoscaling";
import {
    AmazonLinuxImage,
    InstanceClass,
    InstanceSize,
    InstanceType,
    ISecurityGroup,
    IVpc,
    SubnetType
} from "@aws-cdk/aws-ec2";
import {ApplicationLoadBalancer} from "@aws-cdk/aws-elasticloadbalancingv2";
import {Role} from "@aws-cdk/aws-iam";

interface ComputeStackProps extends cdk.StackProps {
    vpc: IVpc;
    instanceRole: Role;
    appSg: ISecurityGroup;
}

export class ComputeStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);


        const prj: string = this.node.tryGetContext('prj');
        const stage: string = this.node.tryGetContext('stage');
        // const params: any = this.node.tryGetContext(stage);

        const asg = new AutoScalingGroup(this, 'ASG', {
            vpc: props.vpc,
            instanceType: InstanceType.of(InstanceClass.BURSTABLE2, InstanceSize.MICRO),
            machineImage: new AmazonLinuxImage(),
            allowAllOutbound: true,
            role: props.instanceRole
        });
        asg.addUserData(
            "yum -y update",
            "yum -y install nginx",
            "systemctl enable nginx",
            "systemctl start nginx"
        );

        const alb = new ApplicationLoadBalancer(this, 'Alb', {
            vpc: props.vpc,
            vpcSubnets: {subnetType: SubnetType.PUBLIC},
            internetFacing: true,
            loadBalancerName: `${prj}-${stage}-alb`
        });

        const listener = alb.addListener('Listener', {
            port: 80,
        });
        listener.addTargets('Target', {
            targets: [asg],
            port: 80
        });

    }
}