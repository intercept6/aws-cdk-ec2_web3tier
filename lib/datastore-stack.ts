// @aws-cdk/aws-rds is a developer preview (public beta) module.
// Releases might lack important features and might have future breaking changes.

import {
  Construct,
  RemovalPolicy,
  Stack,
  StackProps,
  Tag
} from "@aws-cdk/core";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  ISecurityGroup,
  IVpc,
  SubnetType
} from "@aws-cdk/aws-ec2";
import { DatabaseCluster, DatabaseClusterEngine } from "@aws-cdk/aws-rds";

interface DataStoreStackProps extends StackProps {
  vpc: IVpc;
  dbSg: ISecurityGroup;
}

export class DataStoreStack extends Stack {
  constructor(scope: Construct, id: string, props: DataStoreStackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: any = this.node.tryGetContext(stage);

    const cluster = new DatabaseCluster(this, "DBCluster", {
      engine: DatabaseClusterEngine.AURORA,
      masterUser: {
        username: "admin"
      },
      removalPolicy: RemovalPolicy.DESTROY,
      instanceProps: {
        instanceType: InstanceType.of(
          InstanceClass.MEMORY5,
          InstanceSize.LARGE
        ),
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED
        },
        vpc: props.vpc,
        securityGroup: props.dbSg
      }
    });
    cluster.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
