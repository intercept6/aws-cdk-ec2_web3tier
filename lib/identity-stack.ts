import { Construct, Stack, StackProps } from "@aws-cdk/core";
import { ManagedPolicy, Role, ServicePrincipal } from "@aws-cdk/aws-iam";

export class IdentityStack extends Stack {
  public readonly instanceRole: Role;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: object = this.node.tryGetContext(stage);

    this.instanceRole = new Role(this, "IamRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      roleName: `${prj}-${stage}-app`,
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2RoleforSSM"
        )
      ]
    });
  }
}
