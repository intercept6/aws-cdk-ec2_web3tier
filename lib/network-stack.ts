import { Construct, Stack, StackProps, Tag } from "@aws-cdk/core";
import { Port, SecurityGroup, SubnetType, Vpc } from "@aws-cdk/aws-ec2";

export class NetworkStack extends Stack {
    public readonly vpc: Vpc;
    public readonly appSg: SecurityGroup;
    public readonly dbSg: SecurityGroup;

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const prj: string = this.node.tryGetContext("prj");
        const stage: string = this.node.tryGetContext("stage");
        const params: any = this.node.tryGetContext(stage);

    this.vpc = new Vpc(this, "VPC", {
      enableDnsHostnames: true,
      enableDnsSupport: true,
      natGateways: params.nat_gateways,
      maxAzs: 2,
      subnetConfiguration: [
        { name: "Public", subnetType: SubnetType.PUBLIC, cidrMask: 24 },
        { name: "Private", subnetType: SubnetType.PRIVATE, cidrMask: 24 },
        { name: "Isolated", subnetType: SubnetType.ISOLATED, cidrMask: 24 }
      ]
    });
    this.vpc.node.applyAspect(new Tag("Name", `${prj}-${stage}-vpc`));

    this.appSg = new SecurityGroup(this, "AppSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-app`,
      description: `${prj}-${stage}-app`
    });
    this.appSg.addIngressRule(
      this.appSg,
      Port.tcp(80),
      "allow http access from alb"
    );
    this.appSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-app`));

    this.dbSg = new SecurityGroup(this, "DbSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-db`,
      description: `${prj}-${stage}-db`
    });
    this.dbSg.addIngressRule(
      this.appSg,
      Port.tcp(3306),
      "allow db access from app server"
    );
    this.dbSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
