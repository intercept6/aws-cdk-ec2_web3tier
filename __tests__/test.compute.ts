import { App, Stack } from "@aws-cdk/core";
import { SynthUtils } from "@aws-cdk/assert";
import { ComputeStack } from "../lib/compute-stack";
import { SecurityGroup, Vpc } from "@aws-cdk/aws-ec2";
import { Role, ServicePrincipal } from "@aws-cdk/aws-iam";

describe("compute", () => {
  test("default", () => {
    const app = new App();
    app.node.setContext("prj", "ssprj");
    app.node.setContext("stage", "ssstage");
    app.node.setContext("ssstage", {
      nat_gateways: 1
    });
    const givenStack = new Stack(app, "GivenStack");
    const instanceRole = new Role(givenStack, "TestRole", {
      roleName: "test",
      assumedBy: new ServicePrincipal("ec2.amazonaws.com")
    });
    const vpc = new Vpc(givenStack, "TestVPC");
    const sg = new SecurityGroup(givenStack, "TestSecurityGroup", {
      vpc: vpc
    });
    const stack = new ComputeStack(app, "TestComputeStack", {
      vpc: vpc,
      instanceRole: instanceRole,
      appSg: sg
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
