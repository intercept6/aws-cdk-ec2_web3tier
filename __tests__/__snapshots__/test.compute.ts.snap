// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compute default 1`] = `
Object {
  "Parameters": Object {
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "ASG46ED3070": Object {
      "Properties": Object {
        "DesiredCapacity": "1",
        "LaunchConfigurationName": Object {
          "Ref": "ASGLaunchConfigC00AF12B",
        },
        "MaxSize": "1",
        "MinSize": "1",
        "Tags": Array [
          Object {
            "Key": "Name",
            "PropagateAtLaunch": true,
            "Value": "TestComputeStack/ASG",
          },
        ],
        "TargetGroupARNs": Array [
          Object {
            "Ref": "AlbListenerTargetGroup34CD4313",
          },
        ],
        "VPCZoneIdentifier": Array [
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCPrivateSubnet1Subnet005D5ABE0341BEED",
          },
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCPrivateSubnet2Subnet45EC7DE34DC431C1",
          },
        ],
      },
      "Type": "AWS::AutoScaling::AutoScalingGroup",
      "UpdatePolicy": Object {
        "AutoScalingScheduledAction": Object {
          "IgnoreUnmodifiedGroupSizeProperties": true,
        },
      },
    },
    "ASGInstanceProfile0A2834D7": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestRole6C9272DF9DD89F08",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "ASGInstanceSecurityGroup0525485D": Object {
      "Properties": Object {
        "GroupDescription": "TestComputeStack/ASG/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestComputeStack/ASG",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCD26570D8E54DE509",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ASGInstanceSecurityGroupfromTestComputeStackAlbSecurityGroupCE0D67B380F4465ED4": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ASGInstanceSecurityGroup0525485D",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "AlbSecurityGroup580F65A6",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ASGLaunchConfigC00AF12B": Object {
      "Properties": Object {
        "IamInstanceProfile": Object {
          "Ref": "ASGInstanceProfile0A2834D7",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t2.micro",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "ASGInstanceSecurityGroup0525485D",
              "GroupId",
            ],
          },
        ],
        "UserData": Object {
          "Fn::Base64": "#!/bin/bash
yum -y update
yum -y install nginx
systemctl enable nginx
systemctl start nginx",
        },
      },
      "Type": "AWS::AutoScaling::LaunchConfiguration",
    },
    "Alb16C2F182": Object {
      "Properties": Object {
        "Name": "ssprj-ssstage-alb",
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "AlbSecurityGroup580F65A6",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCPublicSubnet1SubnetAE333572D53FE905",
          },
          Object {
            "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCPublicSubnet2SubnetF2FC694369D844D7",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "AlbListener86261768": Object {
      "Properties": Object {
        "DefaultActions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "AlbListenerTargetGroup34CD4313",
            },
            "Type": "forward",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "Alb16C2F182",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "AlbListenerTargetGroup34CD4313": Object {
      "Properties": Object {
        "Port": 80,
        "Protocol": "HTTP",
        "TargetType": "instance",
        "VpcId": Object {
          "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCD26570D8E54DE509",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "AlbSecurityGroup580F65A6": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB TestComputeStackAlb356BBA99",
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "GivenStack:ExportsOutputRefTestVPCD26570D8E54DE509",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "AlbSecurityGrouptoTestComputeStackASGInstanceSecurityGroupAA82C57980B75EECA4": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ASGInstanceSecurityGroup0525485D",
            "GroupId",
          ],
        },
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "AlbSecurityGroup580F65A6",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
  },
}
`;
